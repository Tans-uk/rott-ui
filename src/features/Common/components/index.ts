export * from './CommonItem'
export * from './CommonItemContainer'
