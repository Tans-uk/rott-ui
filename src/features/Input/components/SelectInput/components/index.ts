export * from './SelectInput'
export * from './SelectInputModalComponent'
