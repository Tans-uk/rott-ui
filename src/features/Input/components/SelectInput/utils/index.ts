export * from './listHeightNormalizer'
export * from './modalHeightPercentageNormalizer'
export * from './sortListBySearchPriority'
