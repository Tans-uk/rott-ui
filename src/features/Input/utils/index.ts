export * from './formatByDateMode'
export * from './inputNormalizer'
