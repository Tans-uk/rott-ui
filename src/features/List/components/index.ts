export * from './List'
export * from './FavoriteSwipeComponent'
export * from './ListSkeletonItem'
