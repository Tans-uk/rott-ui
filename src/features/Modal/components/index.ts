export * from './ModalComponent'
export * from './ModalContainer'
