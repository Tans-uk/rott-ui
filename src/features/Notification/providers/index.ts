export * from './ToasterNotificationProvider'
export * from './NotificationProvider'
