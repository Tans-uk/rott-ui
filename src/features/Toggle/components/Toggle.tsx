import {useEffect, useMemo, type FC} from 'react'

import {ToggleStyles} from '../style'

import {Item} from '@features/Item'
import {Pressable} from '@features/Pressable'
import {themeConfig} from '@providers'
import {display} from '@utils'

import Animated, {
  Easing,
  interpolate,
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated'

interface ToggleProps {
  testID?: string
  isOn?: boolean
  onToggleChange?: (checked: boolean) => void
  disabled?: boolean
}

export const Toggle: FC<ToggleProps> = ({testID, isOn, onToggleChange, disabled}) => {
  const animatedValue = useSharedValue(isOn === true ? 1 : 0)
  const marginRange = useMemo(() => [display.px(4), display.px(36)], [])

  const moveToggleStyle = useAnimatedStyle(
    () => ({
      marginLeft: interpolate(animatedValue.value, [0, 1], marginRange, 'clamp'),
    }),
    []
  )

  const moveToggleBackgroundColorStyle = useAnimatedStyle(
    () => ({
      backgroundColor:
        animatedValue.value === 1 ? themeConfig.colors.primary : themeConfig.colors['grey-200'],
    }),
    []
  )

  const animateToggle = (isChecked: boolean) => {
    animatedValue.value = withTiming(isChecked ? 1 : 0, {
      duration: 100,
      easing: Easing.linear,
    })
  }

  const handleToggle = (isChecked: boolean) => {
    animateToggle(isChecked)
    onToggleChange && runOnJS(onToggleChange)(isChecked)
  }

  useEffect(() => {
    if (isOn === undefined) return
    animateToggle(isOn)
  }, [isOn])

  return (
    <Item testID={testID} row alignItemsCenter width={64} height={32} maxWidth={64} maxHeight={32}>
      <Pressable
        onPress={() => !disabled && handleToggle!(!isOn)}
        testID='toggle-container-test-id'>
        <Animated.View
          testID='toggle-test-id'
          style={[ToggleStyles().toggleContainer, moveToggleBackgroundColorStyle]}>
          <Animated.View style={[ToggleStyles().toggleWheelStyle, moveToggleStyle]} />
        </Animated.View>
      </Pressable>
    </Item>
  )
}
