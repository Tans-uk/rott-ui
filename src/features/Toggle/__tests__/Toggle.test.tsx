import React from 'react'
import {fireEvent, render, waitFor} from '../../../__tests__/utils/testUtils'
import {themeConfig} from '../../../providers'
import {Toggle} from '../components'

describe('Toggle -> Custom Component', () => {
  const toggleContainerTestId = 'toggle-container-test-id'
  const toggleTestId = 'toggle-test-id'

  it('toggle ilk render anında snapshot ile eşleşmeli', () => {
    const renderedToggle = render(<Toggle isOn={false} />)

    expect(renderedToggle).toMatchSnapshot()
  })

  it('toggle default propslar ile render olmalı', () => {
    const {getByTestId} = render(<Toggle isOn={false} />)

    const toggleContainer = getByTestId(toggleContainerTestId)
    expect(toggleContainer).toBeDefined()

    const toggleWheel = getByTestId(toggleTestId)
    expect(toggleWheel).toBeDefined()
  })

  it('tıklandığında onToggleChange çağırılmalı', async () => {
    const onToggleMock = jest.fn()
    const {getByTestId} = render(<Toggle onToggleChange={onToggleMock} isOn={false} />)

    const toggleContainer = getByTestId(toggleContainerTestId)
    fireEvent.press(toggleContainer)

    await waitFor(() => expect(onToggleMock).toHaveBeenCalled())
  })

  describe('isOn özelliği değiştiğinde stil özellikleri değişmeli', () => {
    const onToggleMock = jest.fn()

    it('pasif ise', () => {
      const {getByTestId} = render(<Toggle onToggleChange={onToggleMock} isOn={false} />)

      const toggleContainer = getByTestId(toggleTestId)

      expect(toggleContainer).toHaveStyle({
        backgroundColor: themeConfig.colors['grey-200'],
      })
    })

    it('aktif ise', () => {
      const {getByTestId} = render(<Toggle onToggleChange={onToggleMock} isOn={true} />)

      const toggleContainer = getByTestId(toggleTestId)

      expect(toggleContainer).toHaveStyle({
        backgroundColor: themeConfig.colors.primary,
      })
    })
  })
})
