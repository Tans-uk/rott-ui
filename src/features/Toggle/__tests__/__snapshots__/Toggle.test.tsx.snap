// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle -> Custom Component toggle ilk render anında snapshot ile eşleşmeli 1`] = `
<View
  flex={1}
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    {
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <KeyboardProvider
    enabled={true}
  >
    <View
      flex={1}
      pointerEvents="box-none"
      style={
        {
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
          }
        }
      >
        <View
          alignItemsCenter={true}
          height={32}
          maxHeight={32}
          maxWidth={64}
          style={
            {
              "alignItems": "center",
              "flex": undefined,
              "flexDirection": "row",
              "height": 62,
              "maxHeight": 62,
              "maxWidth": 123,
              "width": 123,
            }
          }
          width={64}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            animated={false}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "flex": 1,
                "flexDirection": "column",
              }
            }
            testID="toggle-container-test-id"
          >
            <View
              collapsable={false}
              jestAnimatedProps={
                {
                  "value": {},
                }
              }
              jestAnimatedStyle={
                {
                  "value": {
                    "backgroundColor": "rgba(161, 173, 175, 1)",
                  },
                }
              }
              jestInlineStyle={
                [
                  {
                    "borderRadius": 16,
                    "height": 62,
                    "justifyContent": "center",
                    "width": 123,
                  },
                ]
              }
              style={
                [
                  {
                    "borderRadius": 16,
                    "height": 62,
                    "justifyContent": "center",
                    "width": 123,
                  },
                  {
                    "backgroundColor": "rgba(161, 173, 175, 1)",
                  },
                ]
              }
              testID="toggle-test-id"
            >
              <View
                collapsable={false}
                jestAnimatedProps={
                  {
                    "value": {},
                  }
                }
                jestAnimatedStyle={
                  {
                    "value": {
                      "marginLeft": 8,
                    },
                  }
                }
                jestInlineStyle={
                  [
                    {
                      "backgroundColor": "rgba(255, 255, 255, 1)",
                      "borderColor": "rgba(34, 63, 70, 0.15)",
                      "borderRadius": 44,
                      "borderWidth": 4,
                      "height": 44,
                      "width": 44,
                    },
                  ]
                }
                style={
                  [
                    {
                      "backgroundColor": "rgba(255, 255, 255, 1)",
                      "borderColor": "rgba(34, 63, 70, 0.15)",
                      "borderRadius": 44,
                      "borderWidth": 4,
                      "height": 44,
                      "width": 44,
                    },
                    {
                      "marginLeft": 8,
                    },
                  ]
                }
              />
            </View>
          </View>
        </View>
        <View
          style={
            {
              "flexShrink": 1,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          [
            {
              "alignSelf": "center",
              "elevation": 999999,
              "flex": 0,
              "maxWidth": "100%",
              "position": "absolute",
              "zIndex": 999999,
            },
            {
              "flexDirection": "column-reverse",
              "justifyContent": "flex-start",
              "top": 30,
              "width": 750,
            },
          ]
        }
      >
        <View
          style={
            {
              "bottom": 0,
            }
          }
        >
          <RCTSafeAreaView />
        </View>
      </View>
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          [
            {
              "alignSelf": "center",
              "elevation": 999999,
              "flex": 0,
              "maxWidth": "100%",
              "position": "absolute",
              "zIndex": 999999,
            },
            {
              "bottom": 40,
              "flexDirection": "column",
              "justifyContent": "flex-end",
              "width": 750,
            },
          ]
        }
      >
        <View
          style={
            {
              "bottom": 0,
            }
          }
        >
          <RCTSafeAreaView />
        </View>
      </View>
    </View>
  </KeyboardProvider>
</View>
`;
